include/flang/lower/OpBuilder.h

```diff
@@ -55,6 +61,107 @@ class SymMap {
   }
 };
 
+/// Helper class that can be inherited from in order to
+/// facilitate mlir::OpBuilder usage.
+class OpBuilderHandler {
+public:
+  OpBuilderHandler(mlir::OpBuilder &b, mlir::Location l) : builder{b}, loc{l} {}
+  template <typename T, typename... Args>
+  auto create(Args... args) {
+    return builder.create<T>(loc, std::forward<Args>(args)...);
+  }
+  mlir::Value getIntegerConstant(mlir::Type integerType, std::int64_t);
+
+protected:
+  mlir::OpBuilder &builder;
+  mlir::Location loc;
+};
+
+/// Facilitate lowering to fir::loop
+class LoopCreator : public OpBuilderHandler {
+public:
+  LoopCreator(mlir::OpBuilder &b, mlir::Location l) : OpBuilderHandler{b, l} {}
+  LoopCreator(OpBuilderHandler &b) : OpBuilderHandler{b} {}
+  // In genLoop functions, lb, ub, and count arguments must have integer types.
+  // They will be automatically converted the IndexType if needed.
+  // Build loop [lb, ub) with step "step".
+  using BodyGenerator = std::function<void(OpBuilderHandler &, mlir::Value)>;
+  void genLoop(mlir::Value lb, mlir::Value ub, mlir::Value step,
+               const BodyGenerator &bodyGenerator);
+  /// Build loop [lb, ub) with step 1.
+  void genLoop(mlir::Value lb, mlir::Value ub,
+               const BodyGenerator &bodyGenerator);
+  /// Build loop [0, count) with step 1.
+  void genLoop(mlir::Value count, const BodyGenerator &bodyGenerator);
+
+private:
+  mlir::Type getIndexType();
+  mlir::Value convertToIndexType(mlir::Value integer);
+};
+
+/// Facilitate lowering of CHARACTER operation
+class CharacterOpsCreator : public OpBuilderHandler {
```

Comment:
I am not sure what is the best name for these classes, I took "OpsCreator", suffix but maybe "Handler" postfix is easier though less precise. Do you have an opinion ?

Reply:
MLIR sometimes refactors classes into Foo and FooImpl.  That might be the right thing here.

Reply:
You mean, you would use `CharacterImpl`, `LoopImpl` and `ComplexImpl` here ?

Reply:
The other direction. Say a `CharacterOpsBuilder` that is derived from `OpBuilderImpl`.

Reply:
OK, thanks for clarifying, I did that except I used `OpBuilderWrapper` instead of `OpBuilderImpl` as to me `OpBuilderImpl` sounds like it would implement the interface of `mlir::OpBuilder`.